### ISSUE TITLE
vmalert: support Grafana Rule Editor via Grafana Ruler API

### ISSUE THREAD

**Issue by lhns** [VictoriaMetrics/VictoriaMetrics#4939]
### Is your feature request related to a problem? Please describe

It is a bit annoying having to edit the rule configuration yaml files by hand. Grafana includes a rule editor that allows you to create and edit rules for specific data sources like mimir, loki or cortex. It would be pretty cool if that would also work for vmalert.

### Describe the solution you'd like

Grafana supports these rule edit features via the [Grafana Ruler API](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#ruler). The API is actually pretty simple as it just receives the yaml from Grafana.
Some time ago I made this proxy for Prometheus to intercept the Ruler API calls, modify the rule yamls and send a reload request to prometheus. Turns out this also works pretty well for victoriametrics and vmalert.
https://github.com/lhns/grafana-ruler-proxy

It would be nice though if this feature was natively supported. Maybe behind a flag. Would something like this be considered out-of-scope for this project?

**Comment by hagen1778** [comment]
> Would something like this be considered out-of-scope for this project?

How does https://github.com/lhns/grafana-ruler-proxy gets the access to Prometheus/vmalert filesystem to make a change?

**Comment by lhns** [comment]
The Prometheus Configs have to be mounted into the ruler proxy container. Not the most elegant solution but it works.

**Comment by hagen1778** [comment]
> The Prometheus Configs have to be mounted into the ruler proxy container. Not the most elegant solution but it works.

vmalert supports loading config rules via HTTP:
```
  -rule array
     Path to the files or http url with alerting and/or recording rules.
     Supports hierarchical patterns and regexpes.
     Examples:
      -rule="/path/to/file". Path to a single file with alerting rules.
      -rule="http://<some-server-addr>/path/to/rules". HTTP URL to a page with alerting rules.
```

Do you think it could be better for proxy to provide an API for fetching the config instead of changing it on the filesystem?

**Comment by lhns** [comment]
I didn't know that. That is a really good idea!

**Comment by huozhirui** [comment]
> > The Prometheus Configs have to be mounted into the ruler proxy container. Not the most elegant solution but it works.
> 
> vmalert supports loading config rules via HTTP:
> 
> ```
>   -rule array
>      Path to the files or http url with alerting and/or recording rules.
>      Supports hierarchical patterns and regexpes.
>      Examples:
>       -rule="/path/to/file". Path to a single file with alerting rules.
>       -rule="http://<some-server-addr>/path/to/rules". HTTP URL to a page with alerting rules.
> ```
> 
> Do you think it could be better for proxy to provide an API for fetching the config instead of changing it on the filesystem?

HTTP method does not support wildcard loading. Will it be supported in the future?
eg: -rule="http://<some-server-addr>/path/to/rules/*.yml

**Comment by hagen1778** [comment]
> HTTP method does not support wildcard loading. Will it be supported in the future?

I don't think it is possible to use wildcards for URLs.
Filesystem API or object storage API provide `List()` method which returns all files listed for the given path. Based on this list and provided wildcard, vmalert can pick files that satisfy the expression.

But with URL - it is not clear how to get the list of files available at `http:///path/to/rules`.  So it can't support widlcards.

**Comment by fayak** [comment]
That would be a very nice feature to have natively implemented in VictoriaMetrics.
I'd like to go for VictoriaMetrics over Mimir for various reasons, but the alerting/notifications part is yet a bit troublesome as Grafana Alerting UI allows to edit the rules and the notifications policies for Mimir but not for Victoriametrics/Alertmanager. The Grafana UI is definitely to be preferred over YAML provisioning I reckon, it's way simpler and suitable for all the people that will be interested in creating alerts without much knowledge in terraform/YAML/...

**Comment by xdr34m** [comment]
Just want to add,  this feature would be insane to have natively.
Im currently running Mimir instead of VMmetrics only to have this feature. I would swap if implemented! :)

**Comment by hagen1778** [comment]
@fayak @xdr34m Grafana already allows defining alerting rules in UI without need to use [Grafana Ruler API](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#ruler) - [Grafana-managed alert rules](https://grafana.com/docs/grafana/latest/alerting/alerting-rules/create-grafana-managed-rule/). This approach is datasource-agnostic, which seems more reasonable for UI alerting.

Supporting  [Grafana Ruler API](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#ruler) could be complicated, because it is specific to Mimir's implementation of ruling. vmalert doesn't have a concept of [ring](https://grafana.com/docs/mimir/latest/references/http-api/#ruler-ring-status), for example.

**Comment by Bruno-DaSilva** [comment]
@hagen1778 I agree re: defining alerts, grafana-managed alert rules have worked extremely well in my use cases at least. 

Where it falls short is with products like Grafana SLO or Grafana K8S Monitoring which explicitly setup recording rules (and i think, a few alerts too) on the _datasource_. There are no Grafana-managed recording rules that work very well.

I agree it would be awesome if there was a component that supported the Ruler API (at least the parts that interact with alert rules -- i don't think the ring status webpage is necessary at all). That would enable these newer cloud services to interact with VictoriaMetrics. It's been a huge bummer for us that we can't use Grafana SLO.

-------------
One idea for implementation for kubernetes deployments of victoriametrics is to have something like @lhns's proxy write to kubernetes configmaps, and then mount those configmaps inside vmalert. Effectively, using k8s configmap(s) as the state store.
But you could implement it to write to s3 or as well, and then load rules in via vmalert's object store reading if you have enterprise.

In either case if this was natively supported in vmalert, the state changes of rule edits would have to be stored somewhere. In a HA deployment of vmalert the local filesystem wouldn't be enough as that isn't shared across multiple vmalert instances.

**Comment by tiny-pangolin** [comment]
One other thing to keep in mind about Grafana managed alerts is that they do not scale very well since the alert state is stored in whatever SQL database Grafana is using by default and storing it in a Prometheus/Victoriametrics or Loki requires extra configuration and making Grafana highly available is more complicated than making vmalert highly available. So, providing a way to mimic the Mimir ruler API would allow users to make alerts in the UI while without having to worry about having to scale Grafana and a SQL database in addition to Victoriametrics. 
Even if the functionality was limited, it would still help users who are migrating off mimir and adopting Victoriametrics. The rules themselves could be stored on shared storage like NFS or object storage that can be accessed and updated by multiple vmalerts

**Comment by andriibeee** [comment]
@hagen1778 @tiny-pangolin @Haleygo Grafana uses a single datasource URL for both queries and ruler calls - it can't be configured to send them to different backends. So either something needs to sit in front and route /config/v1/rules/... to vmalert and everything else to VictoriaMetrics or we can just put the ruler endpoints in vmselect/vmsingle. Wdyt?

**Comment by tiny-pangolin** [comment]
Easiest thing for users would be to have vmselect/vmsingle proxy these endpoints similar to vmalert.proxyURL. not every user uses vmauth and maintaining docs for how to route requests and supporting users to configure vmauth or another http proxy would be challenging if we relied on using an http proxy to route requests

**Comment by andriibeee** [comment]
@tiny-pangolin i think we can simply rely on vmalert.proxyURL

**Comment by Haleygo** [comment]
We can proxy all `config/v1/rules` requests to `vmalert.proxyURL`, the same way we currently handle `/api/v1/rules`.

Please note that this feature is currently under evaluation and there is no guarantee it will be included in the final product.

**Comment by tiny-pangolin** [comment]
Adding this feature might also unlock the ability for users that want to use Victoriametrics with Grafana cloud SLOs with Victoriametrics since this feature depends on being able to manage rules using this API
